prov/gni/src/gnix_nic.c

```diff
@@ -52,12 +52,231 @@ static pthread_mutex_t gnix_nic_list_lock = PTHREAD_MUTEX_INITIALIZER;
+	unsigned int recov;
+
+	/*
+	 * first see if there are any CQE's on the
+	 * tx cq. Calling GNI_CqTestEvent does not require
+	 * getting the nic lock
+	 */
+
+	status = GNI_CqTestEvent(nic->tx_cq);
+	if (status == GNI_RC_NOT_DONE)
+		return FI_SUCCESS;
+
+try_again:
+	fastlock_acquire(&nic->lock);
+        status = GNI_CqGetEvent(nic->tx_cq, &cqe);
+        if (status  == GNI_RC_NOT_DONE) {
+		fastlock_release(&nic->lock);
+		return FI_SUCCESS;
+	}
+
+	switch (status) {
+	case GNI_RC_SUCCESS:
+		assert(GNI_CQ_STATUS_OK(cqe));
+		/*
+		 * check whether CQE from SMSG or a Post
+		 * transaction
+		 */
+		if (GNI_CQ_GET_TYPE(cqe)
+				== GNI_CQ_EVENT_TYPE_POST) {
+			status2 = GNI_GetCompleted(nic->tx_cq, cqe, &gni_desc);
+			if ((status2 != GNI_RC_SUCCESS) &&
+				(status2 != GNI_RC_TRANSACTION_ERROR)) {
+				ret = gnixu_to_fi_errno(status2);
+			}
+			gnix_tdesc = container_of(gni_desc,
+						struct gnix_tx_descriptor,
+						desc.gni_desc);
+		}  else if (GNI_CQ_GET_TYPE(cqe)
+			    == GNI_CQ_EVENT_TYPE_SMSG) {
+			msg_id = GNI_CQ_GET_MSG_ID(cqe);
+			gnix_tdesc = gnix_desc_lkup_by_id(nic,msg_id);
+			if (gnix_tdesc == NULL)
+				ret = -FI_ENOENT;
+		} else {
+			assert(0);   /* TODO: something better -unexpected event type */
+		}
+
+		fastlock_release(&nic->lock);
+		if (ret == FI_SUCCESS) {
+			if (gnix_tdesc->desc.completer_func) {
+				ret =
+				   gnix_tdesc->desc.completer_func(gnix_tdesc);
+				if (ret)
+					goto err;
+                        }
+                        ret = _gnix_nic_tx_free(nic, gnix_tdesc);
+			if (ret)
+				goto err;
+		}
+		break;
+
+	case GNI_RC_TRANSACTION_ERROR: /* uh oh, a hiccup in the network,
+					stupid user, etc. */
+		/* this shouldn't happen SMSG */
+		if (GNI_CQ_GET_TYPE(cqe) == GNI_CQ_EVENT_TYPE_SMSG) {
+			char ebuf[512];
+			GNI_CqErrorStr(cqe, ebuf, sizeof(ebuf));
+			GNIX_WARN(FI_LOG_EP_DATA,
+				  "CQ error statusfor GNI_SmsgSend - %s\n",
+				  ebuf);
+			goto err1;
+		}
+		status2 = GNI_GetCompleted(nic->tx_cq, cqe, &gni_desc);
+		fastlock_release(&nic->lock);
+		gnix_tdesc = container_of(gni_desc,
+					struct gnix_tx_descriptor,
+					desc.gni_desc);
+		if ((status2 != GNI_RC_SUCCESS) &&
+			(status2 != GNI_RC_TRANSACTION_ERROR)) {
+			ret = gnixu_to_fi_errno(status2);
+			goto err;
+		}
+		/*
+ 		 * TODO: need to allow for recover of failed transactions
+ 		 */
+		status = GNI_CqErrorRecoverable(cqe,&recov);
+		if ((status == GNI_RC_SUCCESS) && !recov) {
+			char ebuf[512];
+			GNI_CqErrorStr(cqe, ebuf, sizeof(ebuf));
+			GNIX_WARN(FI_LOG_EP_DATA,
+				  "CQ error statusfor GNI_Post - %s\n",
+				  ebuf);
+			goto err;
+		}
+
+		break;
+
+	default:
+		assert(0);  /* TODO: better error later */
+	}
+
+	/*
+	 * keep on dequeuing until we get GNI_RC_NOT_DONE
+	 */
+
+	goto try_again;
+err1:
+	fastlock_release(&nic->lock);
+err:
+	return ret;
+
+}
+
+/*
+ * process an RX CQEs for this nic
+ */
+static int __nic_rx_progress(struct gnix_nic *nic)
+{
+	return -FI_ENOSYS;
```

Comment:
Maybe this should just return FI_SUCCESS, until it's implemented so that `_gnix_nic_progress` succeeds.